shared/ssz/hash_cache.go

```diff
@@ -0,0 +1,234 @@
+package ssz
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/karlseguin/ccache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+var (
+	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root.
+	ErrNotMerkleRoot = errors.New("object is not a merkle root")
+	// maxCacheSize is 2x of the follow distance for additional cache padding.
+	// Requests should be only accessing blocks within recent blocks within the
+	// Eth1FollowDistance.
+	maxCacheSize = params.BeaconConfig().HashCacheSize
+	// Metrics
+	hashCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "hash_cache_miss",
+		Help: "The number of hash requests that aren't present in the cache.",
+	})
+	hashCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "hash_cache_hit",
+		Help: "The number of hash requests that are present in the cache.",
+	})
+	hashCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "hash_cache_size",
+		Help: "The number of hashes in the block cache",
+	})
+)
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *ccache.Cache
+}
+
+// root specifies the hash of data in a struct
+type root struct {
+	Hash       common.Hash
+	MerkleRoot []byte
+}
+
+// newHashCache creates a new hash cache for storing/accessing root hashes from
+// memory.
+func newHashCache() *hashCacheS {
+	return &hashCacheS{
+		hashCache: ccache.New(ccache.Configure().MaxSize(maxCacheSize)),
+	}
+}
+
+// RootByEncodedHash fetches Root by the encoded hash of the object. Returns true with a
+// reference to the root if exists. Otherwise returns false, nil.
+func (b *hashCacheS) RootByEncodedHash(h common.Hash) (bool, *root, error) {
+	item := b.hashCache.Get(h.Hex())
+	if item == nil {
+		hashCacheMiss.Inc()
+		return false, nil, nil
+	}
+	hashCacheHit.Inc()
+	hInfo, ok := item.Value().(*root)
+	if !ok {
+		return false, nil, ErrNotMerkleRoot
+	}
+
+	return true, hInfo, nil
+}
+
+// TrieRootCached computes a trie root and add it to the cache.
+// if the encoded hash of the object is in cache, it will be retrieved from cache.
+// This method also trims the least recently added root info. if the cache size
+// has reached the max cache size limit.
+func (b *hashCacheS) TrieRootCached(val interface{}) ([32]byte, error) {
+	if val == nil {
+		return [32]byte{}, newHashError("untyped nil is not supported", nil)
+	}
+	rval := reflect.ValueOf(val)
+	hs, err := hashedEncoding(rval)
+	if err != nil {
+		return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
```

Comment:
cant we just use `err.Error()` and get the string ? no need to use `fmt.Sprint`

Reply:
done